pag-lexer/src/normalization.rs

```diff
@@ -6,182 +6,193 @@
-                return ns;
-            }
-            // r. epsilon = r
-            if matches!(&*ns, Epsilon) {
-                return nr;
-            }
-            // epsilon . s = s
-            if matches!(&*nr, Epsilon) {
-                return ns;
-            }
-            if let Concat(r1, r2) = &*nr {
-                return normalize(Rc::new(Concat(r1.clone(), Rc::new(Concat(r2.clone(), ns)))));
-            }
+recursive_flatten! {
+    flatten_union, Union
+}
 
-            if Rc::ptr_eq(&nr, r) && Rc::ptr_eq(&ns, s) {
-                return tree;
+recursive_flatten! {
+    flatten_intersection, Intersection
+}
+
+fn sequence_unchanged(x: &[Rc<RegexTree>], y: &[Rc<RegexTree>]) -> bool {
+    x.iter()
+        .map(|x| x.as_ref() as *const RegexTree)
+        .eq(y.iter().map(|y| y.as_ref() as *const RegexTree))
+}
+
+pub fn normalize(node: Rc<RegexTree>) -> Rc<RegexTree> {
+    match node.as_ref() {
+        Bottom | Epsilon | Set(..) => node,
+        Concat(old) => {
+            let new: RcVec = old.iter().map(|x| normalize(x.clone())).collect();
+            let new: RcVec = flatten_concat(&new);
+            if new.iter().any(|x| matches!(x.as_ref(), Bottom)) {
+                return RegexTree::bottom();
+            }
+            let new: RcVec = new
+                .into_iter()
+                .filter(|x| !matches!(x.as_ref(), Epsilon))
+                .collect();
+            if new.len() == 1 {
+                new[0].clone()
+            } else if new.is_empty() {
+                RegexTree::epsilon()
+            } else if sequence_unchanged(&new, old) {
+                node
+            } else {
+                Rc::new(Concat(new))
             }
-            Rc::new(Concat(nr, ns))
         }
-        KleeneClosure(r) => {
-            let nr = normalize(r.clone());
-            match &*nr {
-                KleeneClosure(_) => nr,
-                Bottom => Rc::new(Epsilon),
-                Epsilon => Rc::new(Epsilon),
+        KleeneClosure(old) => {
+            let new = normalize(old.clone());
+            match &*new {
+                KleeneClosure(_) => new,
+                Bottom | Epsilon => RegexTree::epsilon(),
                 _ => {
-                    if Rc::ptr_eq(r, &nr) {
-                        return tree;
+                    if Rc::ptr_eq(old, &new) {
+                        node
+                    } else {
+                        Rc::new(KleeneClosure(new))
                     }
-                    Rc::new(KleeneClosure(nr))
                 }
             }
         }
-        Union(r, s) => {
-            // referential equality
-            if Rc::ptr_eq(r, s) {
-                return normalize(r.clone());
-            }
-            let r = normalize(r.clone());
-            let s = normalize(s.clone());
+        Union(old) => {
+            let new: RcVec = old.iter().map(|x| normalize(x.clone())).collect();
+            let new: RcVec = flatten_union(&new);
```

Comment:
Subtrees should have already been flattened before this?

Reply:
addressed